Dexie-cloud-addon doesn't get along with vitejs

I'm trying to update my dexie-cloud backed app to use vitejs to take advantage of that sweet HMR response time. However, it's not getting along well with dexie-cloud.

I created a repro here: https://github.com/dusty-phillips/dexie-with-vite-repro, but will summarize in issue as well since I will probably delete that repo someday in a cleaning sweep:

```
yarn create vite dexie-vite-repro --template react-ts
cd dexie-vite-repro
```

Create a dexie-cloud instance using the normal documentation.

Run `yarn` and `yarn dev` to see the error message:

```
error when starting dev server:
Error: Build failed with 1 error:
node_modules/dexie-cloud-addon/dist/modern/dexie-cloud-addon.js:19:16: error: No matching export in "node_modules/dexie/dist/dexie.mjs" for import "cmp"
    at failureErrorWithLog (/Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:1493:15)
    at /Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:1151:28
    at runOnEndCallbacks (/Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:941:63)
    at buildResponseToResult (/Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:1149:7)
    at /Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:1258:14
    at /Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:629:9
    at handleIncomingPacket (/Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:726:9)
    at Socket.readFromStdout (/Users/dustyphillips/Desktop/Code/dexie-vite-repro/node_modules/esbuild/lib/main.js:596:7)
    at Socket.emit (node:events:390:28)
    at addChunk (node:internal/streams/readable:324:12)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Note that I used Dexie 3.2.1-beta.2, as the error message is reminiscent of the two PRs mentioned in the release notes for it. I wondered if whatever the fix was for that release is also required for dexie cloud addon, but I haven't verified that.

--- a/src/classes/dexie/dexie.ts
+++ b/src/classes/dexie/dexie.ts
@@ -216,7 +216,12 @@ export class Dexie implements IDexie {
     this.use(observabilityMiddleware);
     this.use(cacheExistingValuesMiddleware);
 
-    this.vip = Object.create(this, {_vip: {value: true}}) as Dexie;
+    // vip is a Dexie instance used in db.on('ready') that don't block until db is ready.
+    const vipDexie = Object.create(this) as Dexie;
+    vipDexie._vip = true;
+    vipDexie.Table = createTableConstructor(vipDexie);
+    vipDexie._allTables = {}; // So that it has it's own tables that refers back to the vip instance.
+    this.vip = vipDexie;
 
     // Call each addon:
     addons.forEach(addon => addon(this));

--- a/src/classes/dexie/generate-middleware-stacks.ts
+++ b/src/classes/dexie/generate-middleware-stacks.ts
@@ -38,7 +38,7 @@ export function generateMiddlewareStacks({_novip: db}: Dexie, tmpTrans: IDBTrans
     if (db.core.schema.tables.some(tbl => tbl.name === tableName)) {
       table.core = db.core.table(tableName);
       if (db[tableName] instanceof db.Table) {
-          db[tableName].core = table.core;
+          db[tableName].core = db.vip[tableName].core = table.core;
       }
     }
   });

--- a/src/classes/version/schema-helpers.ts
+++ b/src/classes/version/schema-helpers.ts
@@ -13,40 +13,45 @@ import { safariMultiStoreFix } from '../../functions/quirks';
 import { createIndexSpec, nameFromKeyPath } from '../../helpers/index-spec';
 import { createTableSchema } from '../../helpers/table-schema';
 import { generateMiddlewareStacks } from '../dexie/generate-middleware-stacks';
+import { Table } from '../table';
 
-export function setApiOnPlace({_novip: db}: Dexie, objs: Object[], tableNames: string[], dbschema: DbSchema) {
+export function setApiOnPlace(objs: Array<Dexie | Transaction>, tableNames: string[], dbschema: DbSchema) {
   tableNames.forEach(tableName => {
     const schema = dbschema[tableName];
     objs.forEach(obj => {
       const propDesc = getPropertyDescriptor(obj, tableName);
-      if (!propDesc || ("value" in propDesc && propDesc.value === undefined)) {
-        // Either the prop is not declared, or it is initialized to undefined.
-        if (obj === db.Transaction.prototype || obj instanceof db.Transaction) {
+      if (!propDesc || ("value" in propDesc && (propDesc.value === undefined || propDesc.value instanceof Table))) {
+        // Either the prop is not declared, or it is initialized to undefined, or is already a Table
+        // on parent prototype prop but we need to set it as own prop (on the vip Dexie instance that derives from real instance)
+        if (obj instanceof Transaction) {
           // obj is a Transaction prototype (or prototype of a subclass to Transaction)
           // Make the API a getter that returns this.table(tableName)
           setProp(obj, tableName, {
-            get(this: Transaction) { return this.table(tableName); },
+            get(this: Transaction | Dexie) { return this.table(tableName); },
             set(value: any) {
               // Issue #1039
               // Let "this.schema = dbschema;" and other props in transaction constructor work even if there's a name collision with the table name.
               defineProperty(this, tableName, {value, writable: true, configurable: true, enumerable: true});
             }
           });
-        } else {
-          // Table will not be bound to a transaction (will use Dexie.currentTransaction)
-          obj[tableName] = new db.Table(tableName, schema);
+        } else if (obj instanceof Dexie) {
+          obj._allTables[tableName] = obj[tableName] = new obj.Table(tableName, schema);
         }
       }
     });
   });
 }
 
-export function removeTablesApi({_novip: db}: Dexie, objs: Object[]) {
+export function removeTablesApi(objs: Array<Dexie | Transaction>) {
   objs.forEach(obj => {
-    for (let key in obj) {
-      if (obj[key] instanceof db.Table) delete obj[key];
-    }
-  });
+    (obj instanceof Dexie ? [obj, obj._allTables] : [obj]).forEach((obj: object) => {
+      for (let key in obj) {
+        if (obj[key] instanceof Table) {
+          delete obj[key];
+        }        
+      }  
+    });
+  })
 }
 
 export function lowerVersionFirst(a: Version, b: Version) {
@@ -142,8 +147,8 @@ export function updateTablesAndIndexes(
         // because when this code runs, there may not be any other code
         // that can access any transaction instance, else than this particular
         // upgrader function.
-        removeTablesApi(db, [db.Transaction.prototype]);
-        setApiOnPlace(db, [db.Transaction.prototype], keys(upgradeSchema), upgradeSchema);
+        removeTablesApi([db.Transaction.prototype]);
+        setApiOnPlace([db.Transaction.prototype], keys(upgradeSchema), upgradeSchema);
         trans.schema = upgradeSchema;
 
         // Support for native async await.
@@ -175,8 +180,8 @@ export function updateTablesAndIndexes(
         deleteRemovedTables(newSchema, idbtrans);
       }
       // Restore the final API
-      removeTablesApi(db, [db.Transaction.prototype]);
-      setApiOnPlace(db, [db.Transaction.prototype], db._storeNames, db._dbSchema);
+      removeTablesApi([db.Transaction.prototype]);
+      setApiOnPlace([db.Transaction.prototype], db._storeNames, db._dbSchema);
       trans.schema = db._dbSchema;
     });
   });
@@ -343,7 +348,7 @@ export function readGlobalSchema({_novip: db}: Dexie, idbdb: IDBDatabase, tmpTra
   db.verno = idbdb.version / 10;
   const globalSchema = db._dbSchema = buildGlobalSchema(db, idbdb, tmpTrans);
   db._storeNames = slice(idbdb.objectStoreNames, 0);
-  setApiOnPlace(db, [db._allTables], keys(globalSchema), globalSchema);
+  setApiOnPlace([db, db.vip], keys(globalSchema), globalSchema);
 }
 
 export function verifyInstalledSchema(db: Dexie, tmpTrans: IDBTransaction): boolean {

--- a/src/classes/version/schema-helpers.ts
+++ b/src/classes/version/schema-helpers.ts
@@ -24,10 +24,11 @@ export function setApiOnPlace(objs: Array<Dexie | Transaction>, tableNames: stri
         // Either the prop is not declared, or it is initialized to undefined, or is already a Table
         // on parent prototype prop but we need to set it as own prop (on the vip Dexie instance that derives from real instance)
         if (obj instanceof Transaction) {
-          // obj is a Transaction prototype (or prototype of a subclass to Transaction)
+          // obj is a Transaction instance or the db.Transaction.prototype
+          // (which prototypally derives from Transaction.prototype, see transaction-contructor.ts)
           // Make the API a getter that returns this.table(tableName)
           setProp(obj, tableName, {
-            get(this: Transaction | Dexie) { return this.table(tableName); },
+            get(this: Transaction) { return this.table(tableName); },
             set(value: any) {
               // Issue #1039
               // Let "this.schema = dbschema;" and other props in transaction constructor work even if there's a name collision with the table name.

--- a/src/classes/version/version-constructor.ts
+++ b/src/classes/version/version-constructor.ts
@@ -23,7 +23,6 @@ export function createVersionConstructor(db: Dexie) {
         version: versionNumber,
         storesSource: null,
         dbschema: {},
-        tables: {},
         contentUpgrade: null
       };
     });

--- a/src/classes/version/version.ts
+++ b/src/classes/version/version.ts
@@ -18,7 +18,6 @@ export class Version implements IVersion {
     version: number,
     storesSource: { [tableName: string]: string | null },
     dbschema: DbSchema,
-    tables: {},
     contentUpgrade: Function | null
   }
 
@@ -38,7 +37,7 @@ export class Version implements IVersion {
   }
 
   stores(stores: { [key: string]: string | null; }): IVersion {
-    const db = this.db;
+    const db = this.db._novip;
     this._cfg.storesSource = this._cfg.storesSource ?
       extend(this._cfg.storesSource, stores) :
       stores;
@@ -55,8 +54,8 @@ export class Version implements IVersion {
     // Update the latest schema to this version
     db._dbSchema = dbschema;
     // Update APIs
-    removeTablesApi(db, [db._allTables, db, db.Transaction.prototype]);
-    setApiOnPlace(db, [db._allTables, db, db.Transaction.prototype, this._cfg.tables], keys(dbschema), dbschema);
+    removeTablesApi([db, db.vip, db.Transaction.prototype]);
+    setApiOnPlace([db, db.vip, db.Transaction.prototype], keys(dbschema), dbschema);
     db._storeNames = keys(dbschema);
     return this;
   }

--- a/src/public/types/dexie.d.ts
+++ b/src/public/types/dexie.d.ts
@@ -26,6 +26,7 @@ export interface Dexie {
   readonly name: string;
   readonly tables: Table[];
   readonly verno: number;
+  /** The 'vip' instance - passed to db.on('ready') and allows usage before db.on('ready') completes. */
   readonly vip: Dexie;
 
   readonly _allTables: { [name: string]: Table<any, IndexableType> };

